pkg/rng/entropy.go

```diff
@@ -0,0 +1,164 @@
+	// EntropyFeedTime sets the loop time for seeding /dev/random by /dev/hwrng
+	// in seconds
+	EntropyFeedTime time.Duration = 10
+	// RandomDevice is the linux random device
+	RandomDevice = "/dev/random"
+	// HwRandomDevice is the linux hw random device
+	HwRandomDevice = "/dev/hwrng"
+)
+
+// Can be extended but keep in mind to priorize
+// more secure random sources like hw random over
+// timer, jitter based mechanisms. Zero is the highest
+// priority.
+// <rng-name> : <priority>
+var trngList = map[string]int{
+	"tpm-rng":        0,
+	"intel-rng":      1,
+	"amd-rng":        1,
+	"timeriomem-rng": 2,
+}
+
+// Searches for available True Random Number Generator
+// inside the kernel api and sets the most secure on if
+// available which seeds /dev/hwrng
+func setAvailableTRNG() (bool, error) {
+	var (
+		currentRNG    string
+		availableRNGs []string
+		rngs          []string
+	)
+
+	currentFileData, err := ioutil.ReadFile(HwRandomCurrentFile)
+	if err != nil {
+		return false, err
+	}
+	currentRNG = string(currentFileData)
+
+	availableFileData, err := ioutil.ReadFile(HwRandomAvailableFile)
+	if err != nil {
+		return false, err
+	}
+	availableRNGs = strings.Split(string(availableFileData), " ")
+
+	for key, value := range trngList {
+		if key == currentRNG && value == 0 {
+			return true, nil
+		}
+	}
+
+	for _, rng := range availableRNGs {
+		for key := range trngList {
+			if rng == key {
+				rngs = append(rngs, key)
+			}
+		}
+	}
+
+	if len(rngs) <= 0 {
+		return false, nil
+	}
+
+	if err = ioutil.WriteFile(HwRandomCurrentFile, []byte(rngs[0]), 0644); err != nil {
+		return false, err
+	}
+
+	return true, nil
+}
+
+// UpdateLinuxRandomness seeds random data from
+// /dev/hwrng into /dev/random based on a timer and
+// the entropy pool size
+// Usage:
+// go UpdateLinuxRandomness()
+func UpdateLinuxRandomness() error {
+	good, err := setAvailableTRNG()
+	if err != nil {
+		return err
+	}
+	if !good {
+		return errors.New("Could not find a good TRNG")
+	}
+
+	randomPoolSizeData, err := ioutil.ReadFile(RandomPoolSizeFile)
+	if err != nil {
+		return err
+	}
+
+	formatted := strings.TrimSuffix(string(randomPoolSizeData), "\n")
+	randomPoolSize, err := strconv.ParseUint(formatted, 10, 32)
+	if err != nil {
+		return err
+	}
+
+	hwRng, err := os.OpenFile(HwRandomDevice, os.O_RDONLY, os.ModeDevice)
+	if err != nil {
+		return err
+	}
+
+	defer hwRng.Close()
+
+	rng, err := os.OpenFile(RandomDevice, os.O_APPEND|os.O_WRONLY, os.ModeDevice)
+	if err != nil {
+		return err
+	}
+
+	defer rng.Close()
+
+	for {
+		randomEntropyAvailableData, err := ioutil.ReadFile(RandomEntropyAvailableFile)
+		if err != nil {
+			return err
+		}
+
+		formatted := strings.TrimSuffix(string(randomEntropyAvailableData), "\n")
+		randomEntropyAvailable, err := strconv.ParseUint(formatted, 10, 32)
+		if err != nil {
+			return err
+		}
+
+		randomBytesNeeded := randomPoolSize - randomEntropyAvailable
```

Comment:
This logic doesn't make sense to me. If I cat some entropy into `/dev/random`, it doesn't get arbitrarily added into the pool, it is mixed into it, and entropy_avail doesn't gain exactly the length of what I cat'd into it, but some amount, often much smaller than the size of what I put into it. It also doesn't hurt to put "too much" into /dev/random.

We should try to read/write a fairly-large static buffer instead of adapting the amount from what is in entropy_avail.
Possibly we can pause seeding if entropy_avail is over some threshold (3kB ?)

Reply:
More better than less but yes we can work with a threshold. I will rework it

Reply:
Done